stm32h747i_disco add ethernet support

KozhinovAlexander opened the pull request:
Add ethernet support to the board

lochej commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.dts:
```diff
@@ -54,3 +54,23 @@
 	status = "okay";
 };
+
+/* From UM2411 Rev 4:
+   With the default setting, the Ethernet feature is not working because of
+   a conflict between ETH_MDC and SAI4_D1 of the MEMs digital microphone */
+&mac {
+	status = "okay";
+	pinctrl-0 = <&eth_ref_clk_pa1
+		     &eth_mdio_pa2
+		     &eth_crs_dv_pa7
+		     &eth_mdc_pc1
+		     &eth_rxd0_pc4
+		     &eth_rxd1_pc5
+		     &eth_tx_en_pg11
+		     &eth_txd0_pg13
+		     &eth_txd1_pg12>;
+};
```
Configuration looks good to me ! 👍 

![image](https://user-images.githubusercontent.com/1338015/97783862-bfb29f00-1b9a-11eb-90cc-20f10717f5d2.png)
![image](https://user-images.githubusercontent.com/1338015/97783781-34d1a480-1b9a-11eb-98fa-c7c8b41eaf15.png)
![image](https://user-images.githubusercontent.com/1338015/97783809-5894ea80-1b9a-11eb-9507-dfc95751e598.png)

lochej commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -63,6 +63,10 @@ The current Zephyr stm32h747i_disco board configuration supports the following h
 +-----------+------------+-------------------------------------+
 | GPIO      | on-chip    | gpio                                |
 +-----------+------------+-------------------------------------+
+| ETHERNET  | on-chip    | ethernet                            |
++-----------+------------+-------------------------------------+
+| RNG       | on-chip    | True Random number generator        |
++-----------+------------+-------------------------------------+
 
```
So yes, this board won't have ETH working until the user changes the SB8 and SB21 configuration. That's not ideal...
We have to maybe add the warning to the doc with an (*) ?

```suggestion
| ETHERNET  | on-chip    | ethernet  (*)                       |
+-----------+------------+-------------------------------------+
| RNG       | on-chip    | True Random number generator        |
+-----------+------------+-------------------------------------+

(*) From UM2411 Rev 4:
   With the default setting, the Ethernet feature is not working because of
   a conflict between ETH_MDC and SAI4_D1 of the MEMs digital microphone.
   Make sure you have SB8 closed and SB28 open to get Ethernet working.
```

Nukersson replied:
Thank you. I am absolutely agree - this belongs to be mentioned in documentation.

gmarull commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -63,6 +63,15 @@ The current Zephyr stm32h747i_disco board configuration supports the following h
 +-----------+------------+-------------------------------------+
+| ETHERNET  | on-chip    | ethernet  (*)                       |
++-----------+------------+-------------------------------------+
+| RNG       | on-chip    | True Random number generator        |
```
Each peripheral enablement should have its own commit.

Nukersson replied:
Splitted as suggested.

gmarull reviewed (changes requested):
Hardware tests are also needed. If tested successfully please report the results.

Commit:
boards: arm: stm32h747i_disco: ethernet

add ethernet driver support

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Commit:
boards: arm: stm32h747i_disco: rng

add rng driver support

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Nukersson commented:
> Hardware tests are also needed. If tested successfully please report the results.

Unfortunatelly I do not have this one board. @erwango @ABOSTM do you have appropriate HW to test this PR? Could you do it please?

ABOSTM commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -63,6 +63,15 @@ The current Zephyr stm32h747i_disco board configuration supports the following h
+| RNG       | on-chip    | True Random number generator        |
++-----------+------------+-------------------------------------+
+
+(*) From UM2411 Rev 4:
+   With the default setting, the Ethernet feature is not working because of
+   a conflict between ETH_MDC and SAI4_D1 of the MEMs digital microphone.
+   Make sure you have SB8 closed and SB21 open to get Ethernet working.
```
As I remember, modifying SB78 and SB21 is necessary, but not sufficient:
* SB21, SB45 and R87 should be opened
* SB22, SB44, SB17 and SB8 should be closed

see: https://os.mbed.com/teams/ST/wiki/DISCO_H747I-modifications-for-Ethernet
Could be interesting to point to this link as there are picture to help finding the required changes.

gmarull replied:
Would be nice if Zephyr docs can provide its own set of pictures.

ABOSTM replied:
Please note that Pictures are mine :smiley: 
And so, those pictures can be reused for zephyr docs.

gmarull replied:
@ABOSTM Great, copyright clearance then :smile:

Nukersson replied:
If @ABOSTM agrees, would be it okay, to put [this URL](https://os.mbed.com/teams/ST/wiki/DISCO_H747I-modifications-for-Ethernet) with very short description to appropriate index.rst file?

gmarull replied:
I vote for making these images part of Zephyr docs.

erwango replied:
I agree that pictures would be required (and available directly rather than a link)

Nukersson replied:
Images and corresponding description were added.

ABOSTM commented on boards/arm/stm32h747i_disco/stm32h747i_disco_m7.dts:
```diff
@@ -54,3 +54,23 @@
+		     &eth_rxd1_pc5
+		     &eth_tx_en_pg11
+		     &eth_txd0_pg13
+		     &eth_txd1_pg12>;
```
I am not ethernet expert, but don't we need PE2 ETH_nINT signal ?

lochej replied:
We actually don't need it, at least for other H7 board with ethernet that's totally fine.

Nukersson replied:
I can confirm @lochej words: other boards also with RMII working fine without `PE2 ETH_nINT` signal.

ABOSTM commented:
> > Hardware tests are also needed. If tested successfully please report the results.
> 
> Unfortunatelly I do not have this one board. @erwango @ABOSTM do you have appropriate HW to test this PR? Could you do it please or point someone willing to test it with the HW?

I have a stm32h747i_disco board with all Hardware modifications for Ethernet (https://os.mbed.com/teams/ST/wiki/DISCO_H747I-modifications-for-Ethernet). So yes I can test it.

erwango commented:
> Unfortunatelly I do not have this one board. @erwango @ABOSTM do you have appropriate HW to test this PR? Could you do it please or point someone willing to test it with the HW?

@Nukersson Thanks for your efforts and contribution. Though, it is recommended that people only add the features they're able to test. We don't necessarily have all the required boards (and not often the bandwidth) to test every feature that people add, so we really rely on people being able to test their contribution, otherwise this couldn't scale.

ABOSTM commented:
I run test `samples/net/dhcpv4_client ` and `samples/net/ipv4_autoconf` but both failed with the same error:
```
*** Booting Zephyr OS build zephyr-v2.4.0-1138-g31d2e9e5e189  ***
[00:00:00.000,000] <err> net_if: There is no network interface to work with!
```
As said @erwango, I could not spend much more time to go further.

lochej commented:
> 
> 
> I run test `samples/net/dhcpv4_client ` and `samples/net/ipv4_autoconf` but both failed with the same error:
> 
> ```
> *** Booting Zephyr OS build zephyr-v2.4.0-1138-g31d2e9e5e189  ***
> [00:00:00.000,000] <err> net_if: There is no network interface to work with!
> ```
> 
> As said @erwango, I could not spend much more time to go further.

You miss this in your stm32h747i_disco/Kconfig.board:
```
if NETWORKING

config NET_L2_ETHERNET
	default y

endif # NETWORKING
```

That's why you have this error 😄

Nukersson commented:
> > Unfortunatelly I do not have this one board. @erwango @ABOSTM do you have appropriate HW to test this PR? Could you do it please or point someone willing to test it with the HW?
> 
> @Nukersson Thanks for your efforts and contribution. Though, it is recommended that people only add the features they're able to test. We don't necessarily have all the required boards (and not often the bandwidth) to test every feature that people add, so we really rely on people being able to test their contribution, otherwise this couldn't scale.

@erwango I understand it. Can we try to push this PR with help of @ABOSTM  forward or should I close it? Int the future I will no more add features I can't directly test.

Nukersson commented:
> > I run test `samples/net/dhcpv4_client ` and `samples/net/ipv4_autoconf` but both failed with the same error:
> > ```
> > *** Booting Zephyr OS build zephyr-v2.4.0-1138-g31d2e9e5e189  ***
> > [00:00:00.000,000] <err> net_if: There is no network interface to work with!
> > ```
> > 
> > 
> > As said @erwango, I could not spend much more time to go further.
> 
> You miss this in your stm32h747i_disco/Kconfig.board:
> 
> ```
> if NETWORKING
> 
> config NET_L2_ETHERNET
> 	default y
> 
> endif # NETWORKING
> ```
> 
> That's why you have this error smile

Fixed now, it was bad rebase from my side. Thank you @lochej!

Nukersson commented:
> I run test `samples/net/dhcpv4_client ` and `samples/net/ipv4_autoconf` but both failed with the same error:
> 
> ```
> *** Booting Zephyr OS build zephyr-v2.4.0-1138-g31d2e9e5e189  ***
> [00:00:00.000,000] <err> net_if: There is no network interface to work with!
> ```
> 
> As said @erwango, I could not spend much more time to go further.

@ABOSTM Would you be able to test this PR on your board once again?

ABOSTM commented:
I tested:
* samples/net/sockets/big_http_download
*samples/net/dhcpv4_client

Both are working fine

erwango reviewed (changes requested):
Thanks for this addition, please add some pics.
Otherwise this is fine to me

Commit:
boards: arm: stm32h747i_disco: doc: images

add images for board ethernet modification
add index.rst Ethernet HW modifications description

Signed-off-by: Alexander Kozhinov <AlexanderKozhinov@yandex.com>

Nukersson commented:
> Thanks for this addition, please add some pics.
> Otherwise this is fine to me

@erwango Pictures were added - please review again

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -110,6 +119,33 @@ Default configuration assigns USART1 and UART8 to the CPU1. The Zephyr console
+
+**Diclaimer:** This section is mostly copy-paste of corresponding
+`DISCO_H747I modifications for Ethernet`_ mbed blog post. The author of this
+article sincerely allowed to use the images and he's knowledge about necessary
```
```suggestion
article sincerely allowed to use the images and his knowledge about necessary
```

Nukersson replied:
Done

erwango commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -110,6 +119,33 @@ Default configuration assigns USART1 and UART8 to the CPU1. The Zephyr console
+article sincerely allowed to use the images and he's knowledge about necessary
+HW modifications to get Ethernet working with this board.
+
+To get Ethernet working following HW modifications may be necessary:
```
```suggestion
To get Ethernet working following HW modifications are required:
```

Nukersson replied:
Done

erwango commented:
@gmarull can you review once again?

MaureenHelm commented on boards/arm/stm32h747i_disco/doc/index.rst:
```diff
@@ -119,6 +119,33 @@ Default configuration assigns USART1 and UART8 to the CPU1. The Zephyr console
+Ethernet
+========
+
+**Diclaimer:** This section is mostly copy-paste of corresponding
```
Disclaimer is misspelled

Nukersson replied:
Thnx. It is fixed now.
